### ISSUE TITLE
interpolation method as an argument

### ISSUE THREAD

**Issue by anirudhkaushik2003** [rohitrango/FireANTs#27]
was just registering masks using AffineRegistration and noticed a colored border

![Image](https://github.com/user-attachments/assets/5572f0da-c923-4874-aff5-37b0df29a9c9)

the issue is that it's using bilinear interpolation by default in the code

```python
moved_image = F.grid_sample(moving_image_blur, coords[..., :-1], mode='bilinear', align_corners=True)
```

`suggestion:` interpolation could be taken as an argument. this would make it possible to use nearest neighbour interpolation to register segmentation masks (issue gets compounded with multi-label masks)

**Comment by anirudhkaushik2003** [comment]
```python
class AffineRegistration(AbstractRegistration):
    """Affine registration class for linear image alignment.

    This class implements affine registration between fixed and moving images using
    gradient descent optimization. Affine transformations are linear transformations that 
    include translation, rotation, scaling, and shearing operations.

    Note about initialization and optimization:
     - All initializations assume the format y = Ax + t (rigid, affine, moments)
     - However, optimization works better with the format y = A(x-c) + c + t'  (where c is the center of the image)
     - therefore, we need to compute t' = t - c + Ac as the learnable parameter if `around_center=True`

    Args:
        scales (List[float]): Downsampling factors for multi-resolution optimization.
            Must be in descending order (e.g. [4,2,1]).
        iterations (List[int]): Number of iterations to perform at each scale.
            Must be same length as scales.
        fixed_images (BatchedImages): Fixed/reference images to register to.
        moving_images (BatchedImages): Moving images to be registered.
        loss_type (str, optional): Similarity metric to use. Defaults to "cc".
        optimizer (str, optional): Optimization algorithm - 'SGD' or 'Adam'. Defaults to 'SGD'.
        optimizer_params (dict, optional): Additional parameters for optimizer. Defaults to {}.
        loss_params (dict, optional): Additional parameters for loss function. Defaults to {}.
        optimizer_lr (float, optional): Learning rate for optimizer. Defaults to 0.1.
        mi_kernel_type (str, optional): Kernel type for MI loss. Defaults to 'b-spline'.
        cc_kernel_type (str, optional): Kernel type for CC loss. Defaults to 'rectangular'.
        cc_kernel_size (int, optional): Kernel size for CC loss. Defaults to 3.
        tolerance (float, optional): Convergence tolerance. Defaults to 1e-6.
        max_tolerance_iters (int, optional): Max iterations for convergence. Defaults to 10.
        init_rigid (Optional[torch.Tensor], optional): Initial affine matrix. Defaults to None.
        custom_loss (nn.Module, optional): Custom loss module. Defaults to None.
        blur (bool, optional): Whether to blur images during downsampling. Defaults to True.
        moved_mask (bool, optional): Whether to mask moved image for loss. Defaults to False.
        loss_device (Optional[str], optional): Device to compute loss on. Defaults to None.
        around_center (bool, optional): Whether to apply affine around the center of the image. Defaults to True.
        interpolation_mode (str, optional): Interpolation mode for final transformation. Defaults to 'bilinear'.
        is_label (bool, optional): Whether images are label maps (segmentations). Defaults to False.
    """

    def __init__(self, scales: List[float], iterations: List[int], 
                fixed_images: BatchedImages, moving_images: BatchedImages,
                loss_type: str = "cc",
                optimizer: str = 'Adam', optimizer_params: dict = {},
                loss_params: dict = {},
                optimizer_lr: float = 3e-2,
                mi_kernel_type: str = 'b-spline', cc_kernel_type: str = 'rectangular',
                cc_kernel_size: int = 3,
                tolerance: float = 1e-6, max_tolerance_iters: int = 10, 
                around_center: bool = True,
                init_rigid: Optional[torch.Tensor] = None,
                custom_loss: nn.Module = None,
                blur: bool = True,
                moved_mask: bool = False,   # mask out moved image for loss
                loss_device: Optional[str] = None,
                interpolation_mode: str = 'bilinear',
                is_label: bool = False, **kwargs,
                ) -> None:

        super().__init__(scales=scales, iterations=iterations, fixed_images=fixed_images, moving_images=moving_images, 
                         loss_type=loss_type, mi_kernel_type=mi_kernel_type, cc_kernel_type=cc_kernel_type, custom_loss=custom_loss, loss_params=loss_params,
                         cc_kernel_size=cc_kernel_size, tolerance=tolerance, max_tolerance_iters=max_tolerance_iters, **kwargs)
        device = self.device
        dims = self.dims
        self.blur = blur
        self.loss_device = loss_device
        self.interpolation_mode = interpolation_mode  # Store requested interpolation mode
        self.is_label = is_label  # Whether these are label images
        
        # Store original moving data for label images
        self.original_moving_data = None
        if self.is_label:
            # Store original data for final transformation
            self.original_moving_data = moving_images().clone()
            # Disable blur for label images
            if self.blur:
                print("Disabling blur for label images")
                self.blur = False
                
        # first three params are so(n) variables, last three are translation
        if init_rigid is not None:
            B, D1, D2 = init_rigid.shape
            if D1 == dims and D2 == dims+1:
                affine = init_rigid
            elif D1 == dims+1 and D2 == dims+1:
                affine = init_rigid[:, :-1, :] + 0
            else:
                raise ValueError(f"init_rigid must have shape [N, {dims}, {dims+1}] or [N, {dims+1}, {dims+1}], got {init_rigid.shape}")
        else:
            affine = torch.eye(dims, dims+1).unsqueeze(0).repeat(self.opt_size, 1, 1).to(device)  # [N, D, D+1]

        # whether to apply affine around the center of the image
        self.around_center = around_center 
        self.center = self.fixed_images.get_torch2phy()[:, :self.dims, -1].contiguous()  # the center of the fixed image is (0, 0, ..., 0) in torch space => A0 + t = t in physical space
        self.center = self.center.to(device)
        if self.around_center:
            # we recalibrate the t' parameter from t 
            transl = affine[:, :self.dims, -1]
            transl = transl - self.center + (affine[:, :self.dims, :self.dims] @ self.center[..., None]).squeeze(-1)
            affine[:, :self.dims, -1] = transl
            affine = affine.detach().contiguous()

        self.affine = nn.Parameter(affine.to(device))  # [N, D]
        print(self.affine)
        self.row = torch.zeros((self.opt_size, 1, dims+1), device=device)   # keep this to append to affine matrix
        self.row[:, 0, -1] = 1.0
        self.moved_mask = moved_mask
        # optimizer
        if optimizer == 'SGD':
            self.optimizer = SGD([self.affine], lr=optimizer_lr, **optimizer_params)
        elif optimizer == 'Adam':
            self.optimizer = Adam([self.affine], lr=optimizer_lr, **optimizer_params)
        else:
            raise ValueError(f"Optimizer {optimizer} not supported")
    
    def get_inverse_warped_coordinates(self, fixed_images: Union[BatchedImages, FakeBatchedImages], moving_images: Union[BatchedImages, FakeBatchedImages], shape=None):
        """Get coordinates for inverse warping (implements abstract method)"""
        pass  # Not needed for our current use case, but required by abstract class

    def save_as_ants_transforms(self, filenames: Union[str, List[str]]):
        ''' 
        Save the registration as ANTs transforms (.mat file)
        '''
        if isinstance(filenames, str):
            filenames = [filenames]

        affine = self.get_affine_matrix(homogenous=False)
        n = affine.shape[0]
        check_and_raise_cond(len(filenames)==1 or len(filenames)==n, "Number of filenames must match the number of transforms")
        check_and_raise_cond(check_correct_ext(filenames, PERMITTED_ANTS_TXT_EXT + PERMITTED_ANTS_MAT_EXT), "File extension must be one of {}".format(PERMITTED_ANTS_TXT_EXT + PERMITTED_ANTS_MAT_EXT))
        filenames = augment_filenames(filenames, n, PERMITTED_ANTS_TXT_EXT + PERMITTED_ANTS_MAT_EXT)

        for i in range(affine.shape[0]):
            mat = affine[i].detach().cpu().numpy().astype(np.float32)
            A = mat[:self.dims, :self.dims]
            t = mat[:self.dims, -1]
            if any_extension(filenames[i], PERMITTED_ANTS_MAT_EXT):
                savemat(filenames[i], {'AffineTransform_float_3_3': mat, 'fixed': np.zeros((self.dims, 1)).astype(np.float32)})
                raise NotImplementedError("This function does not work with ANTs mat files")
            else:
                savetxt(filenames[i], A, t)
            logger.info(f"Saved transform to {filenames[i]}")

    def get_affine_matrix(self, homogenous=True):
        """Get the current affine transformation matrix.

        Always get it in the format y = Ax + t

        Args:
            homogenous (bool, optional): Whether to return homogeneous coordinates. 
                Defaults to True.

        Returns:
            torch.Tensor: Affine transformation matrix.
                If homogenous=True: shape [N, D+1, D+1]
                If homogenous=False: shape [N, D, D+1]
        """
        affine = self.affine.clone()
        if self.around_center:  # we need to convert t' to t
            A = affine[:, :self.dims, :self.dims] + 0
            t = affine[:, :self.dims, -1] + 0
            t = t + self.center - (A @ self.center[..., None]).squeeze(-1)
            affine = torch.cat([A, t[..., None]], dim=-1)
        return torch.cat([affine, self.row], dim=1) if homogenous else affine

    
    def get_warped_coordinates(self, fixed_images: Union[BatchedImages, FakeBatchedImages], moving_images: Union[BatchedImages, FakeBatchedImages], shape=None):
        """Get transformed coordinates for warping the moving image.

        Computes the coordinate transformation from fixed to moving image space
        using the current affine parameters.

        Args:
            fixed_images (BatchedImages): Fixed reference images
            moving_images (BatchedImages): Moving images to be transformed
            shape (Optional[tuple]): Output shape for coordinate grid. 
                Defaults to fixed image shape.

        Returns:
            torch.Tensor: Transformed coordinates in normalized [-1,1] space
                Shape: [N, H, W, [D], dims]
        """
        # get coordinates of shape fixed image
        fixed_t2p = fixed_images.get_torch2phy()
        moving_p2t = moving_images.get_phy2torch()
        affinemat = self.get_affine_matrix()
        if shape is None:
            shape = fixed_images.shape
        init_grid = torch.eye(self.dims, self.dims+1).to(fixed_images.device).unsqueeze(0).repeat(affinemat.shape[0], 1, 1)  # [N, dims, dims+1]
        fixed_image_coords = F.affine_grid(init_grid, shape, align_corners=True)  # [N, H, W, [D], dims+1]
        coords = torch.cat([fixed_image_coords, torch.ones(list(fixed_image_coords.shape[:-1]) + [1], device=fixed_image_coords.device)], dim=-1)
        coords = torch.einsum('ntd, n...d->n...t', moving_p2t @ affinemat @ fixed_t2p, coords)  # [N, H, W, [D], dims+1]  
        return coords[..., :-1]

    def optimize(self, save_transformed=False):
        """Optimize the affine registration parameters.

        Performs multi-resolution optimization of the affine transformation
        parameters using the configured similarity metric and optimizer.

        For label images (is_label=True), follows ANTs approach:
        - Uses bilinear interpolation during optimization for stable gradients
        - Applies nearest neighbor only for final transformed output

        Args:
            save_transformed (bool, optional): Whether to save transformed images
                at each scale. Defaults to False.

        Returns:
            Optional[List[torch.Tensor]]: If save_transformed=True, returns list of
                transformed images at each scale. Otherwise returns None.
        """
        verbose = self.progress_bar
        fixed_arrays = self.fixed_images()
        moving_arrays = self.moving_images()
        fixed_t2p = self.fixed_images.get_torch2phy()
        moving_p2t = self.moving_images.get_phy2torch()
        fixed_size = fixed_arrays.shape[2:]
        # save initial affine transform to initialize grid 
        init_grid = torch.eye(self.dims, self.dims+1).to(self.fixed_images.device).unsqueeze(0).repeat(self.opt_size, 1, 1)  # [N, dims, dims+1]

        if save_transformed:
            transformed_images = []

        for scale, iters in zip(self.scales, self.iterations):
            self.convergence_monitor.reset()
            prev_loss = np.inf
            # downsample fixed array and retrieve coords
            size_down = [max(int(s / scale), MIN_IMG_SIZE) for s in fixed_size]

            ## create fixed downsampled image, and blurred moving image
            if self.blur and scale > 1:
                sigmas = 0.5 * torch.tensor([sz/szdown for sz, szdown in zip(fixed_size, size_down)], device=fixed_arrays.device)
                gaussians = [gaussian_1d(s, truncated=2) for s in sigmas]
                fixed_image_down = downsample(fixed_arrays, size=size_down, mode=self.fixed_images.interpolate_mode, gaussians=gaussians)
                moving_image_blur = separable_filtering(moving_arrays, gaussians)
            else:
                # Handle different interpolation modes correctly
                try:
                    fixed_image_down = F.interpolate(fixed_arrays, size=size_down, mode=self.fixed_images.interpolate_mode, align_corners=True)
                except ValueError:
                    # For 'nearest' mode which doesn't support align_corners
                    fixed_image_down = F.interpolate(fixed_arrays, size=size_down, mode=self.fixed_images.interpolate_mode)
                moving_image_blur = moving_arrays

            fixed_image_coords_homo = F.affine_grid(init_grid, fixed_image_down.shape, align_corners=True)  # [N, H, W, [D], dims+1]
            fixed_image_coords_homo = torch.cat([fixed_image_coords_homo, torch.ones(list(fixed_image_coords_homo.shape[:-1]) + [1], device=fixed_image_coords_homo.device)], dim=-1)
            fixed_image_coords_homo = torch.einsum('ntd, n...d->n...t', fixed_t2p, fixed_image_coords_homo)  # [N, H, W, [D], dims+1]  
            
            pbar = tqdm(range(iters)) if verbose else range(iters)
            torch.cuda.empty_cache()
            for i in pbar:
                self.optimizer.zero_grad()
                affinemat = self.get_affine_matrix()
                coords = torch.einsum('ntd, n...d->n...t', moving_p2t @ affinemat, fixed_image_coords_homo)  # [N, H, W, [D], dims+1]
                
                # During optimization, always use bilinear interpolation for stable gradients
                # regardless of the requested interpolation_mode
                try:
                    moved_image = F.grid_sample(moving_image_blur, coords[..., :-1], mode='bilinear', align_corners=True)
                except ValueError:
                    # Fallback for older PyTorch versions if needed
                    moved_image = F.grid_sample(moving_image_blur, coords[..., :-1], mode='bilinear')

                if self.moved_mask:
                    try:
                        moved_mask = F.grid_sample(torch.ones_like(moving_image_blur), coords[..., :-1], mode='nearest', align_corners=True)
                    except ValueError:
                        moved_mask = F.grid_sample(torch.ones_like(moving_image_blur), coords[..., :-1], mode='nearest')
                else:
                    moved_mask = None

                if self.loss_device is not None:
                    moved_image = moved_image.to(self.loss_device)
                    fixed_image_down = fixed_image_down.to(self.loss_device)
                    if moved_mask is not None:
                        moved_mask = moved_mask.to(self.loss_device)
                
                # calculate loss function
                loss = self.loss_fn(moved_image, fixed_image_down, mask=moved_mask)
                loss.backward()
                self.optimizer.step()
                
                # check for convergence
                cur_loss = loss.item()
                if self.convergence_monitor.converged(cur_loss):
                    break
                prev_loss = cur_loss
                if verbose:
                    pbar.set_description("scale: {}, iter: {}/{}, loss: {:4f}".format(scale, i, iters, prev_loss))
            
            if save_transformed:
                # For label images, apply the desired interpolation mode for the final result
                if self.is_label and self.interpolation_mode == 'nearest' and self.original_moving_data is not None:
                    try:
                        final_image = F.grid_sample(
                            self.original_moving_data, 
                            coords[..., :-1], 
                            mode='nearest'
                        )
                    except ValueError:
                        final_image = F.grid_sample(
                            self.original_moving_data, 
                            coords[..., :-1], 
                            mode='nearest',
                            align_corners=None
                        )
                    transformed_images.append(final_image)
                else:
                    transformed_images.append(moved_image)

        if save_transformed:
            return transformed_images
```

this implementation worked for me.

![Image](https://github.com/user-attachments/assets/bddcb9e0-5ff7-484c-a3a8-10fd77a24ec5)

**Comment by rohitrango** [comment]
you can use 'reg.evaluate' to deform any auxiliary data (label maps, etc). I would recommend loading the label map using 'Image.load_file' with the segmentation parameter set to True. This will create a multi channel probabilistic mask which you can resample using 'reg.evaluate' and take the argmax along the channel dimension to find the best label. This works better than using nearest neighbor sampling (and the latter is also not differentiable).

**Comment by anirudhkaushik2003** [comment]
this code handles the differentiability issue the same way Ants does.

**Comment by rohitrango** [comment]
I do not understand. 

what is the issue with bilinear sampling + argmax along channel dimension?
